### PR TITLE
Refuse a generic type argument the walker cannot name, instead of guessing (E622)

### PR DESCRIPTION
**Pull request by aallan**
## What

Monomorphization infers each of a generic's type arguments by naming the call argument that fixes it, with a hand-maintained walker (`Monomorphizer._infer_vera_type_name`). When no arm matched the argument's shape, the type variable stayed unbound and was **silently substituted with the phantom-variable default `Bool`**.

This PR makes that substitution fail closed where it is a guess, and keeps it where it is sound.

## Root cause

There are two independently maintained walkers that must agree on a clone's name: discovery's `_infer_vera_type_name` (`vera/monomorphize.py`) and the WASM call-site rewrite's `_infer_vera_type` (`vera/wasm/inference.py`). Neither reads the checker's typed artifacts. When their arm sets differ, discovery registers one specialisation and the rewrite calls another — and the phantom default made that disagreement **silent**: check-green, verify-clean, caller dropped at codegen ([E602] cascading to [E620]) or an invalid module at load.

The discriminator this PR adds is the one the family actually turns on:

- A type variable **no parameter position determines** (`E` in `result_unwrap_or(Ok(x), d)`) is not inferable from arguments by construction, and the emitted WASM is identical whatever it is named. The default is kept.
- A type variable a parameter binds **directly** (`fn idg(@T -> @T)`) is determined by that argument's type. Arriving at the default means no arm named the argument, so `Bool` is a guess. That is now the registered **`E622`**.

## What changed

- **New diagnostic `E622`** — "Cannot infer a generic call's type argument" — registered in `vera/errors.py` and `vera/_since.py`, with rationale, fix and `spec_ref` per `check_diagnostic_fields.py`. It names the type variable, the callee, and the AST shape whose arm is missing, and is located on the argument rather than the enclosing function.
- **Discovery** records the un-inferable case (`UninferredTypeArg`) instead of guessing silently; the default is still applied, so no downstream table changes shape.
- **Codegen** drains the records after discovery and reports `E622` as an error — the instantiation set is what it emits clones from.
- **The verifier** drains them too. Its discovered set is what the #732 emitted-versus-discovered differential holds codegen to, and a set built on the guess reports a tier for a specialisation codegen never emits while leaving the real one unverified.
- **The call-site rewrite** answers `None` ("inference failed", its documented contract) rather than mangling a symbol on a guess.
- Records from the **shadowed/qualified** discovery's per-call throwaway walkers are carried out to the same drain, so `mod$plib$gen2$Int` refuses where the unshadowed spelling does.
- **A `ResultRef` arm on the discovery side** — see "New bug found" below.

## Evidence

Measured on `release/v0.2.0` (`6dc41d40`) and on this branch, in separate checkouts with their own venvs (canary asserted both sides).

| Issue | base | this branch |
|---|---|---|
| [#1327](https://github.com/aallan/vera/issues/1327) (`IndexExpr` argument) | check OK, verify OK, `run` → `[E602] call target 'idg$Int' not registered`, `main` dropped | `[E622]` at compile **and** at verify, on the argument |
| [#1366](https://github.com/aallan/vera/issues/1366) (nested `ModuleCall` argument) | check OK, verify OK, `run` → `[E602] call target 'gen2$Int' not registered` | `[E622]` at compile **and** at verify |
| #1366 shadowed spelling | `run` → `[E602] call target 'mod$plib$gen2$Int' not registered` | `[E622]` at compile |
| [#1365](https://github.com/aallan/vera/issues/1365) (type-changing pipe) | check OK, verify OK, load fails `expected i64, found i32` | **unchanged, and deliberately so** |
| [#1357](https://github.com/aallan/vera/issues/1357) (piped module generic) | `[E602]` naming the bare `idg`, `main` dropped via `[E620]` | unchanged |

`#1365` is the boundary of a fail-closed default and is pinned as such: both walkers name the pipe `Int` from its LEFT operand, so nothing is unbound and no default is reached. A confidently wrong answer is invisible here; it needs the type to come from the checker instead. `#1357` does reach a default, but not through an un-nameable direct argument: the piped call never reaches the type-argument inference with the piped operand at all, so `mod$plib$idg$Bool` is discovered from an empty argument list while the rewrite emits no clone name whatsoever. Both are PR-A2's work.

`tests/test_uninferred_type_arg_e622.py` (16 tests) builds each issue's own repro verbatim, asserts both sides of the discriminator, and pins the two boundaries. Mutation-validated one edit at a time — dropping the record fails 6 cells, recording every unbound variable 2 (the genuine phantom), dropping the codegen drain 5, the verifier drain 2, the shadowed-walk threading 1, the rewrite-side refusal 1, demoting the diagnostic to a warning 4, and reverting the `ResultRef` arm 2.

## Blast radius

Base versus head over all **287 corpus programs** (244 conformance + 43 examples), two independent instruments:

- **Diagnostics + WAT**: `check` / `verify` / `compile` outcome, the diagnostic-code set, and a path-normalised SHA of the emitted WAT — **0 programs moved**. 250 check-ok, 249 verify-ok, 248 compile-ok on both sides.
- **Obligations**: every program's `verify --json` summary, per-status counts and diagnostic/warning codes — **0 programs moved**; 2,181 `verified` + 806 `tier3` + 1 `violated` on both sides, totals identical.

So no currently-green program relied on the phantom substitution — the S4 gate this PR had to clear before the default could be made loud.

Reach per changed arm: the fail-closed record fires on no corpus program (its reach is the four repro fixtures and the suite); the `ResultRef` arm changes no corpus WAT either — its reach is `tests/test_generic_where_helper_990.py` and the new cells, which is exactly why the gap had gone unnoticed.

## New bug found

Making the default loud turned a silent guess into a refusal, and the refusal landed on `tests/test_generic_where_helper_990.py::test_contract_only_instantiation_runs`. Root-caused: **discovery has no `ResultRef` arm at all**, while the WASM twin has read one since #912. Measured on `6dc41d40`, unmodified:

- A `forall<T>` `where` helper reached ONLY through the postcondition — `ensures(gok(@Int.result))` — is `vera check`-green, `vera verify`-green, and `vera run` reports `[E602] call target 'parent$where$gok$Int' not registered in this module`, dropping `main` via `[E620]`.
- The same helper also called from `requires(gok(@Int.0))` runs correctly — but only by coincidence: the slot-reference argument in the `requires` registers the `$Int` clone the postcondition's call needs. Its WAT carries **both** `$parent$where$gok$Bool` (the guess) and `$parent$where$gok$Int`.

That shape wants the arm rather than a refusal — the walker can name a `ResultRef` exactly as its twin does — so `ResultRef` now shares the `SlotRef` arm on both sides. After the change the `ensures`-only program runs (15) and the coincidence variant emits only `$parent$where$gok$Int`. Tracked as [#1369](https://github.com/aallan/vera/issues/1369). Its KNOWN_ISSUES / ROADMAP rows are deliberately absent, because this PR fixes it: `check_doc_counts.py --check-bug-issues`'s single divergence is then exactly this PR's fixed set.

## The net's boundary (measured, not assumed)

Widening the fail-closed record to a variable reached only through a parameter's TYPE ARGUMENTS was implemented and run. It **refuses the prelude's own combinators**: `result_unwrap_or(result_map(Ok(100), f), 0)` has an argument no arm can name, so `VeraE` — reachable only through `Result<T, E>` — is recorded and reported, taking `ch09_prelude`, `ch09_option_result_combinators`, both dual-target cells and four unit tests red. That variable is exactly the phantom the default documents, and the prelude stands on it.

So the net covers the case where a wrong name is a wrong WASM **type** — a bare `@T` parameter, whose clone takes the argument's own representation — and leaves the case where the variable is a type argument of a heap constructor the clone handles as a pointer. The exclusion is stated for what it is rather than claimed sound: `fn head(@Array<T> -> @T)` returns AT `T`, so a guessed `head$Bool` would return i32 where `Int` needs i64, and this net does not catch it. Four cells pin the boundary in both directions — the nested shape is not refused, its clone carries the guessed name, the nameable control is named correctly, and the prelude shape the widening would have refused still compiles and runs.

The gap closes by **inference, not refusal**: once the argument's type comes from the checker there is nothing left to guess, and the same change leaves the prelude working because inference succeeds there too. Measured on the PR-A2 branch, the reviewer's fixture `takes_arr(@Array<Array<Int>>.0[0])` emits `takes_arr$Int` on both consultors and runs, where this branch emits `takes_arr$Bool`.

## Bookkeeping

- `KNOWN_ISSUES.md` gains Bugs rows for **#1365** and **#1366**, which had none; the burndown table in `ROADMAP.md` gains both beside #1327, and its header count moves to twenty-two. Verified against `gh issue list --label bug --state open` (22).
- `CHANGELOG.md` `[Unreleased]` gains a Changed bullet (the fail-closed step, which makes #1327/#1366 loud without closing them) and a Fixed bullet for #1369.
- `check_doc_counts.py --check-bug-issues` shows six divergences at this head: **#1369** (this PR's fix, correctly untabled), **#1322** (#1370's fix, already merged), and **#1371 / #1374 / #1375 / #1376** (other streams' new bugs, theirs to table). On a release branch issues close only at the final PR, so an issue fixed here stays open and untabled — the divergence being exactly the fixed set is the intended shape.
- Header word re-derived at the current head per S11: 21 KNOWN_ISSUES Bugs rows == 21 ROADMAP burndown rows == "Twenty-one" (it was "Twenty-two" before #1370 merged and removed #1322's row).
- Doc-count mirrors lockstepped (12,306 tests / 182 files / 161 error codes), `TESTING.md` table row added, site assets regenerated.

## Gates

Full suite `12,107 passed, 173 skipped, 26 deselected`; `mypy vera/` clean; `ruff check` and `ruff check --select S` clean; `check_doc_counts` (plain and `--check-bug-issues`) consistent; `check_limitations_sync` both modes consistent; `check_site_assets` up to date; `check_diagnostic_fields` OK; `check_explicit_encoding` OK; `check_changelog_updated` OK; 244 conformance programs pass; 43 examples pass; `check_examples_run` trap-free; 294 corpus programs canonical; `gitleaks` no leaks. Pre-commit hooks do not run in a worktree without `.venv`, so the gates were run manually and the commit used `--no-verify`.

Fixes #1369 — closes at the v0.2.0 release PR.

Does not close #1327 / #1366 — it makes their failure loud and located. PR-A2 single-sources both walkers from the checker's typed artifacts and closes them along with #1365 and #1357.

## Summary by CodeRabbit

* **New Features**
  * Added E622 diagnostics when generic type arguments cannot be inferred, including actionable source details.
  * Added support for type inference involving result references.

* **Bug Fixes**
  * Improved consistency across compilation, code generation and verification when inference fails.
  * Improved handling of nested constructor tuples and JSON diagnostic envelopes.
  * Preserved valid phantom-variable defaults.

* **Tests**
  * Added 27 regression tests covering E622 diagnostics and generic inference edge cases.

* **Documentation**
  * Updated project status, test counts, error-code references, roadmap and known-issues documentation.

### CODE REVIEW

**Comment by aallan** [comment]
@coderabbitai review

File: KNOWN_ISSUES.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ Defects in shipped compiler, runtime, or tooling behaviour — this table matche
...
...ra check`- and `vera verify`-green and fails at load with `type mismatch: expected i64, found i32`; the `Float64` stage fails the same way with `found f64`, and the direct twin `gid(to_b(@Int.0))` is correct.  Root cause measured: the `BinaryExpr` arm of BOTH type namers — `_infer_vera_type_name` (`vera/monomorphize.py`) and its twin `_infer_vera_type` (`vera/wasm/inference.py`) — returns the LEFT operand's type, which for a `PIPE` is the piped-IN value and is right only where the stage preserves the type; discovery and the call-site rewrite therefore agree confidently on the wrong instantiation, so nothing is unbound and no diagnostic fires at any stage before `wasmtime` refuses the bytes.  Pre-existing — identical before and after PR #1364, which inherits the arm unchanged.  Same check-green / verify-clean family as [#1357](https://github.com/aallan/vera/issues/1357), with an invalid module rather than a dropped caller as the outcome.  Fix direction: both arms must infer the type of the pipe's RESULT — the right-hand call's return type after the left operand is folded into its argument list — so a chain types stage by stage. | [#1365](https://github.com/aallan/vera/issues/1365) |
```
[the remark this answers is not in the document]
Comment (aallan) [reply KNOWN_ISSUES.md]: @coderabbitai Escaped in e39fbc55 — `\|>` in both occurrences, matching the #1357 row's precedent. Confirmed the row renders as two cells again and `check_limitations_sync --check-states` reads its Issue column (73 limitations, consistent). Swept the #1366 row and every other row this PR wrote: no other unescaped pipe.

File: tests/test_uninferred_type_arg_e622.py [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+        desc = errs[0][1]
+        assert "'T'" in desc and "'idg'" in desc and "IndexExpr" in desc, desc
+
+    def test_compile_does_not_emit_the_guessed_clone(self) -> None:
+        # Before the fix the module carried `idg$Bool` (discovery's guess) and
+        # a call to `idg$Int` (the rewrite's answer).  Neither may survive a
+        # refusal: an emitted clone would mean the guess still reached WAT.
+        result = _compile(_INDEX_ARG)
+        assert "idg$Bool" not in result.wat
+        assert "$idg$Int" not in result.wat
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_uninferred_type_arg_e622.py]: @coderabbitai Fixed in e39fbc55. Both clone assertions now use `wat_fn_names` / `wat_calls` as exact sets, with the observed set in the failure message; the new nested-var cells were written that way from the start. You were right about the prefix: `idg$Bool` matches `idg$BoolBox`, which is exactly how one clone impersonates another.

File: tests/test_uninferred_type_arg_e622.py [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+
+    def test_shadowed_qualified_spelling_reports_e622(
+        self, tmp_path: Path,
+    ) -> None:
+        # Discovered by the throwaway-walker path, so this pins that the
+        # record reaches the drain rather than dying with the walker.
+        _verify_errors, _result, cg_errors = build_multi_module(
+            tmp_path / "shadowed", _MODULE_NESTED_SHADOWED,
+        )
+        assert "E622" in [c for c, _ in cg_errors], cg_errors
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_uninferred_type_arg_e622.py]: @coderabbitai Fixed in e39fbc55 — the cell now asserts `verify_errors` carries E622 as well as `cg_errors`. It is the only cell that reaches the verifier's drain through the throwaway-walker path, so leaving it to the sibling cell left that path unasserted.

File: vera/codegen/monomorphize.py [hunk 4]
```diff
@@ -1116,7 +1171,14 @@ def _mono_infer_shadowed(
...
         if op_result_types:
             m._op_result_types = op_result_types
-        return m._infer_type_args_from_args(decl, args, ctor_to_adt, None)
+        result = m._infer_type_args_from_args(decl, args, ctor_to_adt, None)
+        # #1327/#1366: this walker is a THROWAWAY built per qualified call, so
+        # its fail-closed records would be dropped on the floor.  Carry them to
+        # the codegen-level accumulator `_monomorphize` drains, or the shadowed
+        # /qualified spelling of a shape (`mod$plib$gen2$Int`) would keep
+        # guessing where the unshadowed one refuses.
+        self._shadowed_uninferred_type_args.extend(m.uninferred_type_args)
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/codegen/monomorphize.py:1224]: @coderabbitai Deduped in e39fbc55, on the same `(callee, variable, span)` key `_record_uninferred_type_arg` uses, in both drains. Reporting honestly: I could not construct a program that produces a duplicate — a two-instantiation shadowed fixture yields one diagnostic with and without the guard — so it is belt to the per-walker braces rather than a fix for an observed duplicate, and the comment says so. The "exactly one diagnostic per argument" cell pins the property wherever it is supplied from.

File: vera/verifier.py [hunk 5]
```diff
@@ -2312,8 +2312,48 @@ def record_nested(clone: ast.FnDecl, base_chain: str) -> None:
...
+        """Turn discovery's un-inferable type arguments into [E622] errors.
+
+        The verifier's leg of the #1327/#1366 fail-closed: codegen refuses to
+        EMIT a specialisation resting on the phantom-var guess, and this
+        refuses to report a TIER for one.  Both consultors run the same shared
+        inference, so both record the same argument.
+        """
+        for rec in mono.uninferred_type_args:
+            self._error(
+                rec.arg,
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/verifier.py]: @coderabbitai Fixed in e39fbc55. The record now carries the module namespace the walk was in, and the verifier reports inside `_declaring_module_scope(rec.origin)` — the scope the file name, source buffer, naming env and fn registry already travel together in. Red-first cell added: an un-nameable argument written inside an imported module must name that module's file and quote its line, which fails without the origin.

**Comment by aallan** [comment]
@coderabbitai The outside-diff finding on `vera/README.md:701` is addressed in e39fbc55. "No valid program ever fails verification" was a claim about the SMT translation — an untranslatable contract degrades to Tier 3 rather than erroring — and it was being read as a blanket guarantee that verification accepts every type-checking program. E622 makes that reading plainly false, so the sentence now states the narrower thing it always meant and names E622 as the refusal that is not a tier: a generic call whose type argument no walker could name has no determined specialisation to verify. Swept SKILL.md and spec/ for the same sentence — no other occurrence.

**Comment by aallan** [comment]
@coderabbitai review

**Comment by aallan** [comment]
**Closure review at `e39fbc55` (fresh worktree, import pin asserted). Verdict: CLOSED on substance.**

Every code-level fix in the round was mutated and the named cell watched: CR5 (`_declaring_module_scope(rec.origin)` → `(None)`) reds exactly `test_module_owned_site_names_the_module_file`; CR3 (drop `_report_uninferred_type_args(mono)`) reds four cells including the shadowed one, so the verifier leg is asserted as claimed; R1 (fix text back to shape-only) reds `test_fix_names_the_argument_type_not_the_type_variable`; disabling the fail-closed record itself reds nine cells; removing `ResultRef` from the shared `SlotRef` arm reds the two #1369 cells. CR2 verified directly: `wat_calls(w, "idg$Bool")` is `False` on a module defining `$idg$BoolBox` where the old substring test was `True`; imports are excluded from `wat_fn_names`; `return_call` is matched. CR1, R3, R5, R6 verified by reading; CR6's sweep claim holds (one hit repo-wide, now narrowed).

**LOW — CR4's dedupe is pinned by nothing.** Stripping both guards (verifier `seen` and discovery `_uninferred_seen`) leaves the full suite green; `test_one_argument_yields_exactly_one_diagnostic` uses a single-level entry-file call where discovery walks the site once, so it pins existence, not dedupe. The author's own reply already says a duplicating shape could not be constructed and calls the guard belt-and-braces; only the reply's final sentence ("pins the property wherever it is supplied from") overclaims. Accepted fix: a correcting reply on that thread, no code change.

**NIT —** `test_applying_the_fix_verbatim_checks_verifies_and_runs` compiles a pre-fixed fixture rather than extracting the fix text, so it is not red-capable for R1; R1 is pinned by the other cell. Worth keeping as proof the suggested shape checks, verifies and runs to 8.

**R2's boundary** is adequately stated (CHANGELOG bullet with the `result_unwrap_or(result_map(Ok(100), f), 0)` measurement and the ten red tests; the returns-at-`T` reasoning beside the decision in `vera/monomorphize.py`) and the four cells pin both directions, including the known-wrong clone name `takes_arr$Bool` so A2's landing is a visible flip.

**Blast radius** reproduced independently over 287 corpus programs at `fdda84a8` vs `e39fbc55` (`check --json`, `verify --json` incl. per-obligation kind+status, path-normalised WAT digest): byte-identical; tier1 2,181 / tier3 806 / violated 1 on both sides. **Totals:** 12,145 passed, 173 skipped; TESTING.md 12,344 across 184 files; `check_doc_counts` agrees; header "Twenty-one" == 21 rows in both tables. #1369 carries no row by the fixed-in-PR convention.

Merge gates outstanding at review time: CodeRabbit's summary still IN PROGRESS on this head (five inline findings, all replied), CI re-run pending on three cells.

File: vera/codegen/monomorphize.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aallan) [reply vera/codegen/monomorphize.py:1224]: @coderabbitai Correcting my earlier reply on this thread, which overclaimed.

I said the "exactly one diagnostic per argument" cell pins the property wherever it is supplied from. It does not. That cell uses a single-level entry-file call, where discovery walks the site once — so it pins that ONE diagnostic exists, not that a second was suppressed. An independent check confirmed it: stripping BOTH guards (the verifier drain's `seen` and the walker's own `_uninferred_seen`) leaves the full 12,145-test suite green.

Stated plainly: the dedupe is defensive. I have no constructible program that produces a duplicate — a two-instantiation shadowed fixture yields one diagnostic with and without it — so nothing in the suite or the corpus currently distinguishes its presence from its absence.

That is not a stable place to leave code. PR-A2 takes it either way: construct the duplicating shape and pin it with a mutation-red cell, or remove both guards as unreachable. No further commit on this branch.

File: vera/monomorphize.py [hunk 6]
```diff
@@ -1772,8 +1930,18 @@ def namespace_scope(
...
+        # where the walk is, not about whether the consumer built visibility
+        # tables.  The fn-name narrowing below stays conditional.
+        saved_path = self._namespace_path
+        self._namespace_path = path
         if self.ctx.namespace_fn_names is None:
-            yield
+            try:
+                yield
+            finally:
+                self._namespace_path = saved_path
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/monomorphize.py]: @coderabbitai Real bug, fixed in d2ef5898. Confirmed exactly as you describe: the visible-tables branch restored `_scope_fn_names` and not `_namespace_path`, so the path stayed pinned after the block. One save and one restore now cover both fields. Pinned by a direct cell on the manager — nested scopes, then the assertion that the path is `None` again after the outermost block — because the leak is a property of the context manager and a program-level cell would only see it through whichever diagnostic happened to be misattributed. Mutation-validated: reverting the restore fails that cell.

File: vera/codegen/monomorphize.py [hunk 7]
```diff
@@ -1116,7 +1190,14 @@ def _mono_infer_shadowed(
...
+        return result
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/codegen/monomorphize.py:1225]: @coderabbitai Real bug, fixed in d2ef5898. The throwaway walker started outside any scope, so every shadowed-call record claimed `origin=None`. `origin` is now a parameter of `_collect_shadowed_qualified_calls` and `_mono_infer_shadowed` and is applied to the throwaway instance, with the three seed sources giving their three different answers: the entry program's declarations `None`, `_imported_fn_decls` its paired `_mp` (previously discarded), and emitted clones `_mono_clone_origins.get(...)`. Red-first cell added on the shadowed spelling — it asserts the diagnostic names `slib.vera`, and fails without the assignment.

File: vera/codegen/monomorphize.py [hunk 8]
```diff
@@ -387,8 +393,76 @@ def _monomorphize(
...
+                    "and specialising at a guessed type would emit a "
+                    "specialisation nothing calls."
+                ),
+                fix=uninferred_type_arg_fix(
+                    rec, self._expr_semantic_types),
+                spec_ref='Chapter 5, Section 5.9 "Generic Functions"',
+                severity="error",
+                error_code="E622",
+            ))
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/codegen/monomorphize.py:474]: @coderabbitai Real bug, fixed in d2ef5898. The drain now enters `_module_source_scope(rec.origin)` around the location resolution. You are right that only half the misattribution was visible: the verifier's leg already entered its own equivalent scope, so the codegen leg was quoting the importer's line while the verifier quoted the module's. The imported-module cell now asserts the CODEGEN diagnostic's file and source line too — reverting the scope fails it.

File: tests/test_uninferred_type_arg_e622.py [hunk 9]
```diff
@@ -0,0 +1,869 @@
...
+with a fallback value").
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+import pytest
+
+from tests.codegen_helpers import _compile, _run
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_uninferred_type_arg_e622.py:54]: @coderabbitai Fixed in d2ef5898. The four codegen cells now compile through `_compile_checked`, which runs `typecheck_with_artifacts` and threads `expr_semantic_types` / `expr_target_types` as the CLI does. That path is load-bearing here rather than cosmetic: the [E622] `fix` reads the checker's table to name the argument's type, so the artefact-free helper produced the shape-only wording and the cell asserting `let @Int =` could not have passed.

File: tests/test_uninferred_type_arg_e622.py [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+        )
+        assert not check_errors, check_errors
+        vres = verify(program, source, file=str(main_path),
+                      resolved_modules=resolved)
+        e622 = [d for d in vres.diagnostics if d.error_code == "E622"]
+        assert len(e622) == 1, [d.description for d in vres.diagnostics]
+        assert e622[0].location.file is not None
+        assert e622[0].location.file.endswith("mlib.vera"), (
+            e622[0].location.file)
+        assert "idg(" in e622[0].source_line, e622[0].source_line
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_uninferred_type_arg_e622.py:758]: @coderabbitai Fixed in d2ef5898. Added `test_module_owned_site_names_the_module_file_at_codegen`, which drives the same fixture through `build_multi_module` and asserts the E622 in `CompileResult.diagnostics` names `mlib.vera` and quotes the `idg(...)` line. It was red before the `_module_source_scope` fix in the same commit, which is exactly the regression you describe.

File: KNOWN_ISSUES.md [hunk 11]
```diff
@@ -8,6 +8,8 @@ Defects in shipped compiler, runtime, or tooling behaviour — this table matche
...
...output; the single-level `plib::gen(get(()))` in the same file is correct.  Root cause measured: discovery's type namer `_infer_vera_type_name` (`vera/monomorphize.py`) has no `ast.ModuleCall` arm while its WASM call-rewrite twin `_infer_vera_type` (`vera/wasm/inference.py`) does, so the inner call in ARGUMENT position falls to the phantom-`Bool` default and registers `gen2$Bool` while the rewrite emits `gen2$Int`.  Reproduces identically in the shadowed/qualified namespace, where the dropped target is `mod$plib$gen2$Int`.  Pre-existing — identical at `6dc41d40` and at PR #1364's head, which fixes the piped single-stage form and does not reach this arm.  The same two-walker disagreement as [#1357](https://github.com/aallan/vera/issues/1357) and [#1327](https://github.com/aallan/vera/issues/1327), one AST node type over, and the sibling of [#1365](https://github.com/aallan/vera/issues/1365) on the `PIPE` arm.  Fix direction: give discovery a `ModuleCall` arm that resolves the callee through the module table and returns its declared return type, and pin the agreement with the cross-walker differential over a nested-call fixture. | [#1366](https://github.com/aallan/vera/issues/1366) |
```
[the remark this answers is not in the document]
Comment (aallan) [reply KNOWN_ISSUES.md]: @coderabbitai Fixed in d2ef5898, and swept the sibling row too. Both the #1366 and #1327 rows said check and verify were clean and the failure appeared only at `vera run`, which described the pre-change behaviour. They now state that the shape is refused at `vera compile` and `vera verify` with [E622] naming the un-nameable argument, keep the missing walker arm as the open root cause, and say plainly that the refusal is not the fix — the argument is still not inferred, so both issues stay open until the type comes from the checker. The #1365 and #1357 rows are unchanged and still accurate: neither reaches the default, so neither is refused.

**Comment by aallan** [comment]
@coderabbitai Confirming the outside-diff finding on `vera/README.md:701` — "no valid program ever fails verification" — is addressed, and is the same finding my earlier PR-level reply covered. It stands at d2ef5898: the sentence now states the claim it was always making about the SMT translation (an untranslatable contract degrades to Tier 3 rather than erroring) and names E622 as the refusal that is not a tier. Swept SKILL.md and spec/ again at this head — no other occurrence of the sentence.

**Comment by aallan** [comment]
@coderabbitai review

**Comment by aallan** [comment]
**Closure delta at `d2ef5898` (fresh worktrees, import pin asserted). Verdict: CLOSED, no findings.**

The three second-round fixes are each pinned exactly as claimed (baseline 27 green): reverting `namespace_scope` to the split save/restore reds `test_namespace_scope_restores_the_path_it_saved`; dropping the throwaway walker's `_namespace_path = origin` reds `test_module_owned_site_names_the_module_file_at_codegen` and `test_shadowed_module_site_names_the_module_file`; dropping `_module_source_scope(rec.origin)` from the codegen drain reds the same two.

Origin threading attacked with two shapes the PR did not carry, each with deliberately different padding so a wrong file yields a wrong or out-of-range line: a two-hop chain (`main` → `mid` → `deep`, un-nameable call in `deep.vera`) reports `deep.vera:23` at both legs and line 23 really is the call; two sibling modules with their own un-nameable calls report `alpha.vera:14` and `beta.vera:36`, each verified against its own file. Under the throwaway-origin mutation every record collapses to `main.vera` with an out-of-range line and an empty source line, so the green result is evidence, not vacuity. Dual-alias imports are not expressible in the grammar.

`_compile_checked` passes `expr_semantic_types` / `expr_target_types` in the order every CLI compile site does (`cli.py` 665, 887, 1027, 1604); the `expr_types=` spelling at `cli.py:341` is the verify call's own parameter. All eleven CodeRabbit replies check against the code; the three Major findings were real and are fixed as the replies say.

Totals: 12,148 passed / 173 skipped; `check_doc_counts` 12,347 / 184; TESTING.md row closes; header "Twenty-one" == 21 rows in both tables. Diagnostic-location differential over 287 corpus programs (code, file basename, line, column, quoted source line): 522 rows byte-identical between `fdda84a8` and `d2ef5898`.
